store/multiversion/store.go

```diff
@@ -0,0 +1,120 @@
+		multiVersionMap: make(map[string]MultiVersionValue),
+	}
+}
+
+// GetLatest implements MultiVersionStore.
+func (s *Store) GetLatest(key []byte) (value MultiVersionValueItem) {
+	s.mtx.RLock()
+	defer s.mtx.RUnlock()
+
+	keyString := string(key)
+	// if the key doesn't exist in the overall map, return nil
+	if _, ok := s.multiVersionMap[keyString]; !ok {
+		return nil
+	}
+	val, found := s.multiVersionMap[keyString].GetLatest()
+	if !found {
+		return nil // this shouldn't be possible
+	}
+	return val
+}
+
+// GetLatestBeforeIndex implements MultiVersionStore.
+func (s *Store) GetLatestBeforeIndex(index int, key []byte) (value MultiVersionValueItem) {
+	s.mtx.RLock()
+	defer s.mtx.RUnlock()
+
+	keyString := string(key)
+	// if the key doesn't exist in the overall map, return nil
+	if _, ok := s.multiVersionMap[keyString]; !ok {
+		return nil
+	}
+	val, found := s.multiVersionMap[keyString].GetLatestBeforeIndex(index)
+	// otherwise, we may have found a value for that key, but its not written before the index passed in
+	if !found {
+		return nil
+	}
+	// found a value prior to the passed in index, return that value (could be estimate OR deleted, but it is a definitive value)
+	return val
+}
+
+// Has implements MultiVersionStore. It checks if the key exists in the multiversion store at or before the specified index.
+func (s *Store) Has(index int, key []byte) bool {
+	s.mtx.RLock()
+	defer s.mtx.RUnlock()
+
+	keyString := string(key)
+	if _, ok := s.multiVersionMap[keyString]; !ok {
+		return false // this is okay because the caller of this will THEN need to access the parent store to verify that the key doesnt exist there
+	}
+	_, found := s.multiVersionMap[keyString].GetLatestBeforeIndex(index)
+	return found
+}
+
+// This function will try to intialize the multiversion item if it doesn't exist for a key specified by byte array
+// NOTE: this should be used within an acquired mutex lock
+func (s *Store) tryInitMultiVersionItem(keyString string) {
+	if _, ok := s.multiVersionMap[keyString]; !ok {
+		multiVersionValue := NewMultiVersionItem()
+		s.multiVersionMap[keyString] = multiVersionValue
+	}
+}
+
+// Set implements MultiVersionStore.
+func (s *Store) Set(index int, incarnation int, key []byte, value []byte) {
+	s.mtx.Lock()
+	defer s.mtx.Unlock()
+
+	keyString := string(key)
+	s.tryInitMultiVersionItem(keyString)
+	s.multiVersionMap[keyString].Set(index, incarnation, value)
+}
+
+// SetEstimate implements MultiVersionStore.
+func (s *Store) SetEstimate(index int, incarnation int, key []byte) {
+	s.mtx.Lock()
+	defer s.mtx.Unlock()
+
+	keyString := string(key)
+	s.tryInitMultiVersionItem(keyString)
+	s.multiVersionMap[keyString].SetEstimate(index, incarnation)
+}
+
+// Delete implements MultiVersionStore.
+func (s *Store) Delete(index int, incarnation int, key []byte) {
+	s.mtx.Lock()
+	defer s.mtx.Unlock()
+
+	keyString := string(key)
+	s.tryInitMultiVersionItem(keyString)
+	s.multiVersionMap[keyString].Delete(index, incarnation)
```

Comment:
no, because even "deletions" are tracked explicitly, as soon as we introduce an item for a key, that key will remain in the multiversion store map for the remainder of the block, because a later transaction may need to access that data. Once all transactions have completed processing, we can discard the multiversion map after committing to store